src/main/java/io/vertx/core/eventbus/impl/EventBusImpl.java

```diff
@@ -193,13 +193,14 @@ public EventBus unregisterCodec(String name) {
 
   @Override
   public <T> EventBus registerDefaultCodec(Class<T> clazz, MessageCodec<T, ?> codec) {
```

Comment:
Do you want to allow <? super T, ?> here in this PR? See #4113

Reply:
Hey @vietj, it is the other way around isn‘t it? Registering a class to be handled by a:

```java
MessageCodec<? super T, ?>
```

codec means: „Class XYZ is handled by a codec that can handle everything UP-TO that class“, so it is an upper-bound, no lowee-bound. The `<? super T, ?>` don‘t make any assumption about what it is converted into, it just makes it clear that a codec maybe can handle to convert *from* more than just the specific class (it can convert the specific class and all it‘s decendents).

You still only check for the specific class T, in finding the right codec, meaning there is still a 1:1 mapping between class and codec only that you don‘t need to have individual codecs build for types that have a common way of getting serialized. If the codec then spits out the type, or sub type in desirilization is part of the second generic argument which is ? even at the moment. So you cannot be sure right now even for default codecs.